blockchain/client.go

```diff
@@ -77,10 +78,105 @@ func (c *Client) StartEventsListening() (func(), <-chan error, error) {
 			case <-done:
 				return
 			case set := <-sub.Chan():
+				c.processSystemEventsStorageChanges(
+					set.Changes,
+					meta,
+					key,
+					set.Block,
+				)
+			}
+		}
+	}()
+
+	once := sync.Once{}
+	c.cancelListening = func() {
+		once.Do(func() {
+			done <- struct{}{}
+			sub.Unsubscribe()
+			c.isListening = 0
+		})
+	}
+
+	return c.cancelListening, c.errsListening, nil
+}
+
+// RegisterEventsListener subscribes given callback to blockchain events. There is a begin parameter which
+// can be used to get events from blocks older than the latest block. If begin is greater than the latest
+// block number, the listener will start from the latest block.
+func (c *Client) RegisterEventsListener(begin types.BlockNumber, callback EventsListener) (context.CancelFunc, error) {
```

Comment:
are we sure that we want to have `begin` per each listener? wouldn't be better to have it client-wide? 

imagine we have current block number in chain 100 but locally last processed block is 20. In the code we have 5 listeners (that are listen for different even types). How it will work in current implementation? will we fetch missed 80 blocks only once, or per each listener? so 80 or 400(80*5)?

Reply:
> How it will work in current implementation? will we fetch missed 80 blocks only once, or per each listener?

This implementation will fetch the same block multiple times, once per each listener. While not optimal, this behavior is the most straight forward and provides independence to each listener requiring no caching or other more complicated technique to mitigate reads duplication.

We may move `begin` parameter to `StartEvetnsListening` method. But there are a couple of not obvious disadvantages of this. Events listener will have to subscribe before the `StartEventsListening` call, otherwise it may miss an event during the time between `StartEventsListening` and `RegisterEventsListener` calls. And events listeners will have to communicate to understand who is the last one to ensure all callbacks already invoked and the last one can safely mark the block as processed.

Looks like it is easier to deduplicate historical blocks queries using some caching strategy than to deal with mentioned disadvantages in case we are expecting multiple listeners to subscribe at nearly the same time. What do you think?

Reply:
for me the approach to have 'begin' per 'subscriber' looks like complication that's not needed at least for now.

The storage node is going to listen state of blockchain from some position (begin) and subscribe for events in different modules. As for me, at least on first glance, single 'begin' and list of event listeners fully cover our needs and doesn't introduces complexity and overheads.

As for me, the fact that in case event listener was registered after 'start events listening' it can miss some events is fine. As additional protection we can forbid to register event listeners after the 'start' method was invoked. It would force clients (in our case storage node) to register event listeners first and then start listening.

Actually, in case some client needs listening from 2 positions, he can create 2 blockchain clients, right? no need to complicate library... WDYT?

Reply:
the listeners shouldn't communicate between each other and coordinate the storing of 'last processed block'. We should implement it somehow separately from listening logic. The first thought in my mind is to provide callback 'onBlockProcessed' that is triggered once all listneres were invoked without errors. Might be not the best solution, open for discussion here and any ideas you have a very appreciated

Reply:
as an option we can start from the simplest approach - blocking iteration assuming that listeners should be async. In case we see a need to launch goroutines and use wait group - we can update it.

So, i would follow Keep It Stupid Simple approach until there is a real need to complicate it.

Reply:
Updated.